Gemfile

```diff
@@ -39,7 +39,14 @@ end
 
 group :test do
   gem "capybara", ">= 2.15"
+  gem "database_cleaner"
+  gem "factory_bot_rails"
+  gem "ffaker"
+  gem "rails-controller-testing"
```

Comment:
các gem này dùng làm j em

Reply:
- database_cleaner: cung cấp cách để clean database giữa mỗi lúc test. Sau khi chạy test, nó sẽ dọn dẹp database về trang thái trước khi chay để đảm bảo database.
-factory_bot_rails: cho phép tạo ra các object cần thiết cho việc test với các giá trị mặc định, kết hợp cùng với dữ liệu giả tạo ra từ faker hoặc ffaker tạo ra các object với gía trị ngẫu nhiên thay vì chỉ sử dụng giá trị mặc định.
- ffaker dùng để tạo dữ liệu giả để phục vụ cho việc viết test, giống faker nhưng hôm đấy em lỡ cài rồi, mà anh Mạnh bảo dùng ffaker cũng được nên em dùng luôn ạ:))
- rails-controller-testing: gem này dùng để test controller và gán controller đang test đấy với template tương ứng. vd: expect(response).to render_template("index"), nó sẽ hiểu là template index của controller đang test.